File: samples/fb-cpp/events.cpp
/*
 *  events.cpp (fb-cpp) — commit-time delivery, rollback, and coalescing,
 *  through fb-cpp's EventListener.
 *
 *  The fb-cpp twin of ../events_demo.cpp.  The OO-API version hand-rolls the
 *  whole client side: isc_event_block for the EPB, an IEventCallback
 *  implementation with manual addRef/release, isc_event_counts for the
 *  delta, and an explicit re-queue after every delivery (events are
 *  one-shot).  EventListener does all of that internally — it consumes the
 *  baseline delivery, computes per-name deltas, re-queues itself, and hands
 *  the callback a vector of {name, count} on a dispatcher thread.  The three
 *  engine semantics being demonstrated are unchanged: rollback swallows
 *  posts, delivery waits for commit, posts coalesce to one delivery with a
 *  count.  See ../../firebird-events.md.
 *
 *  Build & run (see ../README.md):
 *      ./build/fbcpp_events [database]
 */

#include "fbcpp_sample.h"
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <mutex>
#include <vector>

using namespace fbcpp;
using namespace fbcpp_sample;

// Collects counts delivered by the EventListener's dispatcher thread.
class Delivered
{
public:
	void add(unsigned n)
	{
		std::lock_guard<std::mutex> g{m};
		count += n;
		cv.notify_all();
	}

	// Wait for a delivery (or timeout) and return the coalesced count.
	unsigned take(unsigned timeoutMs)
	{
		std::unique_lock<std::mutex> l{m};
		cv.wait_for(l, std::chrono::milliseconds(timeoutMs), [this] { return count != 0; });
		const unsigned got = count;
		count = 0;
		return got;
	}

private:
	std::mutex m;
	std::condition_variable cv;
	unsigned count = 0;
};

int main(int argc, char** argv)
{
	const char* database = argOrDefault(argc, argv, 1,
		"inet://localhost//tmp/fbhandson/events_fbcpp.fdb");
	int rc = 0;

	try
	{
		Client client{"fbclient"};

		// Two attachments: a listener and a poster.
		Attachment listenAtt = attachOrCreate(client, database);
		Attachment postAtt{client, database, defaultOptions()};

		Delivered delivered;
		EventListener listener{listenAtt, {"demo_event"},
			[&delivered](const std::vector<EventCount>& counts)
			{
				for (const auto& c : counts)
					delivered.add(c.count);
			}};
		printf("listener registered for 'demo_event' (baseline consumed by EventListener)\n");

		// 1. POST_EVENT then ROLLBACK: nothing may be delivered.
		{
			Transaction tra{postAtt};
			postAtt.execute(tra, "execute block as begin post_event 'demo_event'; end");
			tra.rollback();

			const unsigned got = delivered.take(1500);
			printf("after POST_EVENT + ROLLBACK: delivered count = %u  %s\n", got,
				got == 0 ? "(correct - rollback swallows posts)" : "(UNEXPECTED)");
			if (got != 0)
				rc = 1;
		}

		// 2. Three POST_EVENTs in one transaction, then COMMIT:
		//    one delivery with count 3.
		{
			Transaction tra{postAtt};
			postAtt.execute(tra, "execute block as begin"
				" post_event 'demo_event'; post_event 'demo_event'; post_event 'demo_event'; end");
			printf("3 x POST_EVENT executed, not yet committed - waiting briefly...\n");

			const unsigned early = delivered.take(1000);
			printf("before COMMIT: delivered count = %u  %s\n", early,
				early == 0 ? "(correct - delivery is commit-time)" : "(UNEXPECTED)");
			if (early != 0)
				rc = 1;

			tra.commit();
			const unsigned got = delivered.take(3000);
			printf("after COMMIT: delivered count = %u  %s\n", got,
				got == 3 ? "(correct - one delivery, count 3)" : "(UNEXPECTED)");
			if (got != 3)
				rc = 1;
		}

		listener.stop();
		printf(rc == 0 ? "PASS\n" : "FAIL\n");
		return rc;
	}
	catch (const std::exception& e)
	{
		return report(e);
	}
}

File: samples/fpc/events.pas
{
  events.pas — Firebird event notification, in Pascal.

  The fbintf twin of ../events_demo.cpp: two attachments to the same
  database over TCP — a LISTENER that registers interest in
  'demo_event' and a POSTER that executes PSQL blocks containing
  POST_EVENT.  Where the C++ twin implements IEventCallback by hand
  (event block, isc_event_counts, requeue), fbintf wraps the whole
  machinery in IEvents: GetEventHandler + AsyncWaitForEvent arm the
  auxiliary connection, the callback fires on fbintf's event thread,
  and ExtractEventCounts computes the delivered deltas — the same
  isc_event_counts arithmetic, pre-packaged.

  The demo shows the three defining semantics of Firebird events:
    1. delivery happens at COMMIT, not when POST_EVENT executes;
    2. a ROLLBACK discards pending posts — nothing is delivered;
    3. multiple posts of one name in one transaction are delivered
       once, with a count.

  See ../../firebird-events.md.

  Build & run (see ../README.md):
      make -C samples/fpc bin/events && samples/fpc/bin/events
}
program events;

{$mode delphi}
{$codepage UTF8}
{$interfaces COM}

uses
  {$IFDEF UNIX} cthreads, {$ENDIF}
  SysUtils, IB, FBHandsOn;

type
  {AsyncWaitForEvent wants a method of object; the handler runs on
   fbintf's event thread, so it only sets a flag the main thread polls.}
  TListener = class
    Fired: boolean;
    procedure OnEvent(Sender: IEvents);
  end;

procedure TListener.OnEvent(Sender: IEvents);
begin
  Fired := true;
end;

var L: TListener;
    LA, PA: IAttachment;
    EH: IEvents;
    Tr: ITransaction;
    got, early: integer;
    rc: integer;

{Poll for the callback: -1 = nothing delivered before the timeout,
 otherwise the delivered count for 'demo_event' (the callback disarms
 the wait, so the caller must rearm afterwards).}
function WaitFired(timeoutMs: integer): integer;
var waited, i, total: integer;
    counts: TEventCounts;
begin
  waited := 0;
  while (not L.Fired) and (waited < timeoutMs) do
  begin
    Sleep(50);
    Inc(waited, 50);
  end;
  if not L.Fired then Exit(-1);
  L.Fired := false;
  total := 0;
  counts := EH.ExtractEventCounts;
  for i := 0 to High(counts) do
    if counts[i].EventName = 'demo_event' then Inc(total, counts[i].Count);
  Result := total;
end;

begin
  rc := 0;
  L := TListener.Create;
  LA := AttachOrCreate(DbConn('events'));
  PA := FirebirdAPI.OpenDatabase(DbConn('events'), DefaultDPB);

  { Register interest, then prove the plumbing with one committed primer
    post.  Deliveries are one-shot: each callback disarms the wait and
    AsyncWaitForEvent must be called again — fbintf's ExtractEventCounts
    keeps the previous counts as the baseline, so only new posts count. }
  EH := LA.GetEventHandler('demo_event');
  EH.AsyncWaitForEvent(L.OnEvent);
  Tr := PA.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  PA.ExecuteSQL(Tr, 'EXECUTE BLOCK AS BEGIN POST_EVENT ''demo_event''; END', []);
  Tr.Commit;
  got := WaitFired(3000);
  if got >= 0 then
    EH.AsyncWaitForEvent(L.OnEvent);   { delivery disarmed the wait: rearm }
  writeln('listener registered for ''demo_event'' (primer post delivered, count ',
    got, ')');

  { -- 1. POST_EVENT then ROLLBACK: nothing may be delivered. }
  Tr := PA.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taRollback);
  PA.ExecuteSQL(Tr, 'EXECUTE BLOCK AS BEGIN POST_EVENT ''demo_event''; END', []);
  Tr.Rollback;
  got := WaitFired(1500);
  if got < 0 then
    writeln('after POST_EVENT + ROLLBACK: delivered count = 0  '
      + '(correct - rollback swallows posts)')
  else
  begin
    writeln('after POST_EVENT + ROLLBACK: delivered count = ', got, '  (UNEXPECTED)');
    rc := 1;
    EH.AsyncWaitForEvent(L.OnEvent);
  end;

  { -- 2. Three POST_EVENTs in one transaction, then COMMIT:
          one delivery with count 3. }
  Tr := PA.StartTransaction([isc_tpb_write, isc_tpb_nowait, isc_tpb_concurrency],
    taCommit);
  PA.ExecuteSQL(Tr, 'EXECUTE BLOCK AS BEGIN POST_EVENT ''demo_event''; '
    + 'POST_EVENT ''demo_event''; POST_EVENT ''demo_event''; END', []);
  writeln('3 x POST_EVENT executed, not yet committed - waiting briefly...');

  early := WaitFired(1000);
  if early < 0 then
    writeln('before COMMIT: delivered count = 0  (correct - delivery is commit-time)')
  else
  begin
    writeln('before COMMIT: delivered count = ', early, '  (UNEXPECTED)');
    rc := 1;
    EH.AsyncWaitForEvent(L.OnEvent);
  end;

  Tr.Commit;
  got := WaitFired(3000);
  if got < 0 then got := 0;
  if got = 3 then
    writeln('after COMMIT: delivered count = ', got, '  (correct - one delivery, count 3)')
  else
  begin
    writeln('after COMMIT: delivered count = ', got, '  (UNEXPECTED)');
    rc := 1;
  end;

  EH.Cancel;
  EH := nil;
  PA.Disconnect;
  LA.Disconnect;
  L.Free;

  if rc = 0 then writeln('PASS') else writeln('FAIL');
  writeln('done.');
  Halt(rc);
end.
